lib/two_factor_authentication/models/two_factor_authenticatable.rb

```diff
@@ -73,8 +74,14 @@ def totp_enabled?
           respond_to?(:otp_secret_key) && !otp_secret_key.nil?
         end
 
+        def confirm_totp_secret(secret, code, options = {})
+          return false unless authenticate_totp(code, {otp_secret_key: secret})
+          self.otp_secret_key = secret
+          true
+        end
+
         def generate_totp_secret
-          self.otp_secret_key = ROTP::Base32.random_base32
+          ROTP::Base32.random_base32
         end
```

Comment:
Hmm, I agree that the two step "generate"  + "verify" is bigger change, and we can probably find a way to make it simple.  However, I'm not sure I really buy into either of your arguments.

(1) I thought we were OK with breaking backwards compat for this change?  We also broke it by adding an argument to the send_two_factor_authentication_code method.   Remember that generate_totp_secret and confirm_totp_secret are only needed if the developer wants to add Google Authentication support.    Developers who send OTPs to their users (via SMS) don't  ever need to call those.

Another issue here is that when the user wants to generate a new secret and already has an existing one.  This means we need to have an "unconfirmed_totp_secret" separate from "totp_secret" and copy one to the other on confirmation.  While this makes sense to email confirmation (which often happens in a separate session) I think it adds unneeded complexity for confirmations that really should be happening in the same session.

(2) There are lot of metrics that would be useful but I don't think this one database column is the way collect them.  For example, I imagine most systems would want to track all 2FA setup operations, not just the last one.  Also, this change leaves it completely up to the application..  you can choose whatever kind of storage you like for the unconfirmed secret.